mqtt/mqtt-bridge/src/bridge.rs

```diff
@@ -1,31 +1,385 @@
+
+        let pub1 = ReceivedPublication {
+            topic_name: "temp/1".to_string(),
+            qos: QoS::AtLeastOnce,
+            retain: true,
+            payload: Bytes::new(),
+            dup: false,
+        };
+
+        handler
+            .handle_event(Event::Publication(pub1))
+            .await
+            .unwrap();
+
+        let loader = handler.inner.loader().await;
+
+        let extracted1 = loader.lock().next().await.unwrap();
+        assert_eq!(extracted1.1.topic_name, "floor/kitchen/temp/1");
+    }
+
+    #[tokio::test]
+    async fn message_handler_saves_message_with_no_forward_mapping() {
+        let batch_size: usize = 5;
+        let settings = Settings::from_file("tests/config.json").unwrap();
+        let connection_settings = settings.upstream().unwrap();
+
+        let topics: Vec<TopicMapper> = connection_settings
+            .subscriptions()
+            .iter()
+            .map(move |sub| TopicMapper {
+                topic_settings: sub.clone(),
+                topic_filter: TopicFilter::from_str(sub.pattern()).unwrap(),
+            })
+            .collect();
+
+        let mut handler = MessageHandler::<InMemoryPersist>::new(topics, batch_size);
+
+        let pub1 = ReceivedPublication {
+            topic_name: "pattern/p1".to_string(),
+            qos: QoS::AtLeastOnce,
+            retain: true,
+            payload: Bytes::new(),
+            dup: false,
+        };
+
+        handler
+            .handle_event(Event::Publication(pub1))
+            .await
+            .unwrap();
+
+        let loader = handler.inner.loader().await;
+
+        let extracted1 = loader.lock().next().await.unwrap();
+        assert_eq!(extracted1.1.topic_name, "pattern/p1");
+    }
+
+    #[tokio::test]
+    async fn message_handler_no_topic_match() {
+        let batch_size: usize = 5;
+        let settings = Settings::from_file("tests/config.json").unwrap();
+        let connection_settings = settings.upstream().unwrap();
+
+        let topics: Vec<TopicMapper> = connection_settings
+            .subscriptions()
+            .iter()
+            .map(move |sub| TopicMapper {
+                topic_settings: sub.clone(),
+                topic_filter: TopicFilter::from_str(sub.pattern()).unwrap(),
+            })
+            .collect();
+
+        let mut handler = MessageHandler::<InMemoryPersist>::new(topics, batch_size);
+
+        let pub1 = ReceivedPublication {
+            topic_name: "local/temp/1".to_string(),
+            qos: QoS::AtLeastOnce,
+            retain: true,
+            payload: Bytes::new(),
+            dup: false,
+        };
+
+        handler
+            .handle_event(Event::Publication(pub1))
+            .await
+            .unwrap();
 
-    async fn connect_to_local(&self) {
-        info!("connecting to local broker");
+        let key1 = Key { offset: 0 };
+        assert_eq!(None, handler.inner.remove(key1).await);
```

Comment:
This is a hard test to write because of the behavior of the persistor. A remove will always fail if the message hasn't been extracted by the loader. Therefore the test isn't checking correctly right now.

If we want to verify that a message doesn't exist, we will have to prove that the loader can't get it. It's hard because the loader is a stream.

Reply:
It is maybe best to use a more roundabout way of solving this. For example, make a second received publication with a valid topic match and verify the loader retrieves this message first. The test will fail if there is a topic match for the first publication (intended non-topic match) because in this scenario the loader would get this message first.

Reply:
the case when there is a message is if the test runs in parallel with other tests or how would there be another message in the loader. I think it should try to get the message and timeout after a while, but I couldn't make that work.

Reply:
made it work with timer